beam-sync.md

```diff
@@ -0,0 +1,187 @@
+Fast Sync, the block starts executing immediately, collecting *only* the state data needed
+to execute & verify the block. Crucially, a background process collects data that is not
+being read by any block. When this background process is complete, then all of Beam Sync is
+complete, and the node transitions to Full Sync.
+
+## How to Beam Sync
+
+There are many possible approaches to Beam Sync. The plan is to evolve Beam Sync through three stages:
+
+- v0: Use existing Fast Sync protocols; works on mainnet today
+- v1: Add new devp2p protocols, to optionally accelerate Beam Sync
+- v2: Add consensus-sensitive witness proofs to block headers, to enable validation of witnesses, and enforce generation
+
+This document will focus on v0.
+
+### Overview
+
+See the path from starting with an empty node to transitioning to a Full Sync:
+
+![Flow chart](https://i.postimg.cc/3x2qtnGL/Flow-chart-v0-2.png)
+
+### Data Faults
+
+When executing the EVM and collecting data just-in-time (JIT-data), the client can detect that some data is either 1) proven to belong to the state root of the previous header, or 2) created during a local execution of the block. If neither of those is true, then a data fault is generated. The fault must describe the hash of the trie node that's missing. The fault can optionally contain the account, storage, or bytecode lookup that triggered the fault. (This will be useful in v1+ of Beam Sync)
+
+### State Retrieval Performance
+
+Naively, when handling a data fault, each `GetNodeData` request can only contain a single trie node. Starting from the empty trie, you can download the root node, but you can't download any other nodes on the path to the account you care about yet, because they are all hash-indexed. The hash of the child is in the body of the parent. So, proving the balance of a single account with an empty trie takes as many requests as there are levels to the trie (say ~7 in mainnet, currently). 
+
+So performance is highly sensitive to latency: round-trip time to your best peer is the dominating factor of how quickly you can execute the EVM with JIT data.
+
+### Lag
+
+It can take longer than a block time to collect all the state needed for a block. So as new blocks arrive,
+the currently executing block can fall further and further behind the tip. Clients typically
+only serve state for recent blocks, configured differently for different clients. A
+common number is ~100 blocks of state retained. If you fall behind this, then missing state
+may no longer be available to you. This can leave you indefinitely stuck, so when we
+lag far enough behind, we pivot.
+
+### Pivoting
+
+Pivoting skips executing the current blocks, and chooses the tip block again. It will
+typically increase the missing state that must be backfilled. It is quite likely that Beam Sync
+v0 will require a few pivots before it exits. As it progresses, and stores more state, pivoting gets less likely.
```

Comment:
The term *exits* here is a bit ambiguous.  Maybe "before it is able to reliably fetch JIT state data fast enough to avoid .... <something minimally verbose but explicit>"

Reply:
Ah yes, by _exit_ I meant to "switch from Beam Sync to Full Sync". Maybe _completes_?